### DISCUSSION TITLE
Gossip does not stream messages in a continuous fashion

### DISCUSSION THREAD

**Discussion by harsh4786** [libp2p/rust-libp2p#4546] [Q&A]
I'm trying to connect to a node and listen to messages, the problem is it just receives a burst of messages and then the messages stop. Why would that be happening? Here's the code I'm trying to run:-
https://github.com/harsh4786/hopchat/blob/master/src/main.rs

**Comment by thomaseizinger** [comment]
With messages you mean gossipsub messages?

At a first glance, your code looks good but I don't see any part that would actually publish to the topic you are subscribed to. If nobody publishes messages, you can't receive anything :)

**Reply by harsh4786** [reply]
The node that I subscribe to is responsible for publishing the messages. I do receive a burst of messages but then I don't get any more, I want the messages to be in a stream so I can process it further.

**Reply by thomaseizinger** [reply]
Can you share a reproducible example with the code of both nodes and a script that connects them?

Are your connections timing out perhaps? Can you add the ping protocol and check whether those pings still happen continously after the gossipsub messages end?

**Reply by harsh4786** [reply]
Yeah i added a ping protocol as you recommended and i do get continuous pings from the node that i connect to

**Reply by thomaseizinger** [reply]
Can you share a re-producible example of the behaviour?

**Reply by harsh4786** [reply]
yeah this is the working example. https://github.com/harsh4786/hopchat

**Reply by thomaseizinger** [reply]
I added a logger to your example and this is what I found in the output:

```
[2023-10-04T04:49:18Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /wormhole/mainnet/2
[2023-10-04T04:49:18Z DEBUG multistream_select::dialer_select] Dialer: Received rejection of protocol: /wormhole/mainnet/2
```

You are not correctly connecting to the kademlia network of wormwole. I did a bit of digging and found the following line of code: 
https://github.com/Capx-Fi/Wormhole/blob/c626e591e87b945824853c27c3fc861c9b1a1def/node/pkg/p2p/p2p.go#L103

Note that this sets a _prefix_. I'd inquire with the wormhole devs to figure out, what exactly their DHT protocol name is.

For further debugging, I can recommend you run with `RUST_LOG=debug`.

Also, you should call `kad.bootstrap()` on an interval an not just once on startup :)

**Reply by harsh4786** [reply]
Thank you so much for digging this deep! 😃 What exactly do you mean by running `kad.bootstrap()` on an interval?

**Reply by harsh4786** [reply]
This is the official wormhole go implementation of p2p:
https://github.com/wormhole-foundation/wormhole/blob/main/node/pkg/p2p/p2p.go#L260

this is the spy node implementation in go which uses the above p2p:
https://github.com/wormhole-foundation/wormhole/blob/main/node/cmd/spy/spy.go#L52

and these are the arguments to the cli, notice the `—network` argument in the docker command:

https://github.com/wormhole-foundation/wormhole/tree/12-19-wormchain_-_contract_deploy_test/relayer/spy_relayer#running-the-whole-stack-for-testing

**Reply by harsh4786** [reply]
How should I call the `kad.bootstrap() ` on an interval, if you could please elaborate on that? 😅

**Reply by thomaseizinger** [reply]
> How should I call the `kad.bootstrap() ` on an interval, if you could please elaborate on that? 😅

Have a timer that fires every couple of minutes and call bootstrap as a result.

**Reply by thomaseizinger** [reply]
> This is the official wormhole go implementation of p2p: https://github.com/wormhole-foundation/wormhole/blob/main/node/pkg/p2p/p2p.go#L260
> 
> this is the spy node implementation in go which uses the above p2p: https://github.com/wormhole-foundation/wormhole/blob/main/node/cmd/spy/spy.go#L52
> 
> and these are the arguments to the cli, notice the `—network` argument in the docker command:
> 
> https://github.com/wormhole-foundation/wormhole/tree/12-19-wormchain_-_contract_deploy_test/relayer/spy_relayer#running-the-whole-stack-for-testing

You'll need to figure out what the _kademlia_ protocol string is. The wormhole devs will know what that is if you ask them :)

**Comment by harsh4786** [comment]
> > How should I call the `kad.bootstrap() ` on an interval, if you could please elaborate on that? 😅
> 
> Have a timer that fires every couple of minutes and call bootstrap as a result.


Ok so i did this, now it does bootstrap at a certain interval but then i still dont get new messages:
https://github.com/harsh4786/hopchat/blob/master/src/main.rs#L153

**Reply by thomaseizinger** [reply]
Because your kademlia protocol is still wrong: https://github.com/harsh4786/hopchat/blob/428ed740aa8cec713ec23a0e49e62100f5699b80/src/main.rs#L67

**Reply by thomaseizinger** [reply]
You are not actually bootstrapping. You only connect to the bootnode, nothing else. Activate the logs and you'll see it yourself :)

**Comment by thomaseizinger** [comment]
I am closing this as resolved because it is a configuration bug.

**Reply by harsh4786** [reply]
okay, thank you so much for all the help, i really appreciate it! Can i chat with you here if i face any further issues on this topic?

**Reply by thomaseizinger** [reply]
Sure!
